File: services/prediction/prediction/data/quality.py
"""Data-quality scoring and the pre-prediction validation gate."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timedelta

import pandas as pd

from ..config import settings


@dataclass
class QualityReport:
    score: int
    total_records: int
    duplicate_count: int
    missing_sessions: int
    future_timestamps: int
    stale_hours: float | None
    warnings: list[str] = field(default_factory=list)


def compute_quality(df: pd.DataFrame, now: datetime | None = None) -> QualityReport:
    now = now or pd.Timestamp.now(tz="UTC").to_pydatetime()
    warnings: list[str] = []

    total = len(df)
    if total == 0:
        return QualityReport(0, 0, 0, 0, 0, None, ["No historical data available."])

    # True duplicates = same Myanmar-local date AND session with identical
    # result. The same number appearing in MORNING and AFTERNOON is normal,
    # not an error.
    duplicates = int(df.duplicated(subset=["date", "session", "number"]).sum())
    future_ts = int((pd.to_datetime(df["ts"], utc=True) > pd.Timestamp(now) + pd.Timedelta(minutes=5)).sum())

    # Missing sessions: group by Myanmar-local date; trading days (Mon-Fri)
    # should contain both sessions.
    dates = pd.to_datetime(df["date"])
    by_day = df.assign(d=dates).groupby("d")["session"].nunique()
    weekday_days = [d for d in by_day.index if d.weekday() < 5]
    missing_sessions = sum(1 for d in weekday_days if by_day[d] < 2)

    latest_ts = pd.to_datetime(df["ts"], utc=True).max().to_pydatetime()
    stale_hours = max(0.0, (now - latest_ts).total_seconds() / 3600.0)

    w_dup = min(1.0, duplicates / max(20, total * 0.02))
    w_missing = min(1.0, missing_sessions / 40)
    w_future = min(1.0, future_ts / 5)
    w_stale = min(1.0, stale_hours / 72)

    score = round(
        100
        * (
            0.35 * (1 - w_dup)
            + 0.25 * (1 - w_missing)
            + 0.15 * (1 - w_future)
            + 0.25 * (1 - w_stale)
        )
    )

    if score < 90:
        warnings.append("Data quality below 90% — treat analysis with caution.")
    if stale_hours > 72:
        warnings.append("Source data appears stale (>72h). Prediction may be outdated.")
    if duplicates > 0:
        warnings.append(f"{duplicates} duplicate date+result rows detected.")

    return QualityReport(score, total, duplicates, missing_sessions, future_ts,
                         round(stale_hours, 1), warnings)


def quality_gate(df: pd.DataFrame, now: datetime | None = None) -> tuple[bool, list[str]]:
    """Hard gate before prediction generation.

    Returns (passed, failure_reasons). When failed, callers must NOT produce a
    normal prediction ("Prediction unavailable because data validation failed.").
    """
    rep = compute_quality(df, now)
    reasons: list[str] = []
    if rep.total_records < settings.tier2_min:
        reasons.append(
            f"Only {rep.total_records} valid records (<{settings.tier2_min} minimum)."
        )
    if rep.future_timestamps > 0:
        reasons.append(f"{rep.future_timestamps} future timestamps found.")
    if settings.strict_validation and rep.stale_hours is not None and rep.stale_hours > 96:
        reasons.append("Source data older than 96 hours.")
    return len(reasons) == 0, reasons


def data_tier(total_records: int) -> str:
    """Cold-start strategy tiers (spec: COLD START STRATEGY)."""
    if total_records < settings.tier2_min:
        return "TIER_1"
    if total_records < settings.tier3_min:
        return "TIER_2"
    if total_records < settings.tier4_min:
        return "TIER_3"
    return "TIER_4"

File: apps/api/src/services/quality.ts
import { config } from "../config";
import { prisma } from "../db";

export interface QualityGate {
  passed: boolean;
  score: number;
  reasons: string[];
}

/**
 * DATA QUALITY SCORE (0-100), built from measurable components:
 *   completeness (expected 2 sessions per trading day over trailing window)
 *   duplicate rate, invalid records, missing sessions, future timestamps,
 *   source freshness.
 * A gate() check must pass before the engine may generate predictions.
 */
export async function computeDataQuality(): Promise<{
  score: number;
  totalRecords: number;
  duplicateCount: number;
  invalidCount: number;
  missingSessions: number;
  futureTimestamps: number;
  staleHours: number | null;
  warnings: string[];
}> {
  const warnings: string[] = [];
  const totalRecords = await prisma.result.count();
  if (totalRecords === 0) {
    return {
      score: 0,
      totalRecords: 0,
      duplicateCount: 0,
      invalidCount: 0,
      missingSessions: 0,
      futureTimestamps: 0,
      staleHours: null,
      warnings: ["No data ingested yet."],
    };
  }

  // Duplicates: same Myanmar-local date AND session with the identical result
  // is a true duplicate; the same number across both sessions is legitimate.
  const dupRows = await prisma.result.groupBy({
    by: ["date", "session", "twod"],
    _count: { _all: true },
    having: { twod: { _count: { gte: 2 } } },
    orderBy: [{ date: "desc" }],
    take: 200,
  });
  const duplicateCount = dupRows.length;

  // Future timestamps (clock/data corruption signal).
  const futureTimestamps = await prisma.result.count({
    where: { sourceTimestamp: { gt: new Date(Date.now() + 5 * 60_000) } },
  });

  // Rejected records from sync logs.
  const rejectedAgg = await prisma.dataSyncLog.aggregate({
    _sum: { recordsRejected: true },
  });
  const invalidCount = rejectedAgg._sum.recordsRejected ?? 0;

  // Missing sessions: trading days in trailing 60d with only one session.
  const since = new Date(Date.now() - 60 * 24 * 3600_000);
  const grouped = await prisma.result.groupBy({
    by: ["date"],
    where: { date: { gte: since }, sourceTimestamp: { lte: new Date() } },
    _count: { session: true },
    orderBy: [{ date: "desc" }],
  });
  let missingSessions = 0;
  for (const g of grouped) {
    if ((g._count.session ?? 0) < 2) missingSessions++;
    const dow = new Date(g.date).getUTCDay();
    if (dow === 0 || dow === 6) missingSessions--; // weekends are not draws
  }
  missingSessions = Math.max(0, missingSessions);

  // Freshness of latest source event.
  const latest = await prisma.result.findFirst({
    orderBy: { sourceTimestamp: "desc" },
    select: { sourceTimestamp: true },
  });
  const staleHours = latest
    ? Math.max(0, (Date.now() - latest.sourceTimestamp.getTime()) / 3600_000)
    : null;

  // Weighted scoring.
  const w = {
    duplicates: Math.min(1, duplicateCount / Math.max(20, totalRecords * 0.02)),
    invalid: Math.min(1, invalidCount / Math.max(50, totalRecords * 0.05)),
    missing: Math.min(1, missingSessions / 40),
    future: Math.min(1, futureTimestamps / 5),
    stale: staleHours === null ? 1 : Math.min(1, staleHours / 72),
  };
  const score = Math.round(
    100 *
      (0.3 * (1 - w.duplicates) +
        0.15 * (1 - w.invalid) +
        0.25 * (1 - w.missing) +
        0.1 * (1 - w.future) +
        0.2 * (1 - w.stale))
  );

  if (score < 90) warnings.push("Data quality below 90% — treat analysis with caution.");
  if (w.stale === 1) warnings.push("Source data appears stale (>72h old).");
  if (missingSessions > 4) warnings.push(`${missingSessions} trading days have missing sessions.`);

  return {
    score,
    totalRecords,
    duplicateCount,
    invalidCount,
    missingSessions,
    futureTimestamps,
    staleHours: staleHours === null ? null : Math.round(staleHours * 10) / 10,
    warnings,
  };
}

/** Hard gate before prediction generation (spec §16 additional constraints). */
export async function qualityGate(): Promise<QualityGate> {
  const q = await computeDataQuality();
  const reasons: string[] = [];
  if (q.totalRecords < 100) reasons.push(`Only ${q.totalRecords} records (<100 minimum).`);
  if (q.futureTimestamps > 0) reasons.push(`${q.futureTimestamps} future timestamps found.`);
  if (config.strictValidation && q.staleHours !== null && q.staleHours > 96)
    reasons.push("Source data older than 96 hours.");
  return { passed: reasons.length === 0, score: q.score, reasons };
}
